includes/class-wc-checkout.php

```diff
@@ -49,19 +53,23 @@ class WC_Checkout {
 	 * @since 2.1
 	 * @static
 	 * @return WC_Checkout Main instance
+	 *
+	 * @deprecated 4.3.0 Use dependency injection instead, see the ObjectContainer class.
 	 */
 	public static function instance() {
-		if ( is_null( self::$instance ) ) {
-			self::$instance = new self();
+		return self::$instance = ObjectContainer::get_instance_of( __CLASS__ );
+	}
 
-			// Hook in actions once.
-			add_action( 'woocommerce_checkout_billing', array( self::$instance, 'checkout_form_billing' ) );
-			add_action( 'woocommerce_checkout_shipping', array( self::$instance, 'checkout_form_shipping' ) );
+	/**
+	 * Class constructor, hooks the appropriate actions.
+	 */
+	public function __construct() {
+		// Hook in actions once.
+		add_action( 'woocommerce_checkout_billing', array( $this, 'checkout_form_billing' ) );
```

Comment:
I wonder if in our refactored work we might not want to place these kinds of hooks in constructors? The container allows for us to call methods when resolving the instances so maybe we should take advantage of that? Having to worry about hooked actions and filters in tests has proven the bane of my existence, haha.

Reply:
I think that the "hook this as soon as the class is created" logic is something that belongs on the class constructor. Using a separate method would make sense if it was something that could happen on different circumstances too, which is not the case.

Reply:
My thinking here is mostly just in relation to unit testing. Imagine that we've got a constructor that does some actual object construction and then after all of that adds some action handlers. Those handlers might be worth seeing as side-effects in our testing and it just feels like something to avoid in an effort to minimize the possible surface in a test. The container allows for calling methods on resolution so perhaps that's a way we might call an `add_actions` or something? 

This is more something to consider for new classes I think though and isn't worth arguing over for the legacy code. What do you think a good paradigm might be to isolate WordPress hooking from our test code unless explicitly desired?

Reply:
> What do you think a good paradigm might be to isolate WordPress hooking from our test code unless explicitly desired?

Yes, and that's precisely one of the reasons for using the `LegacyProxy` class. If any action/filter registration is made via this class, for tests we can use a custom class that skips hooks as needed, regardless of whether that registration is done in the constructor or not.